staging/src/github.com/openshift/oc/pkg/cli/version/version.go

```diff
@@ -1,25 +1,199 @@
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	kversion "k8s.io/kubernetes/pkg/kubectl/cmd/version"
+	"k8s.io/kubernetes/pkg/kubectl/util/i18n"
+	"k8s.io/kubernetes/pkg/kubectl/util/templates"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	"github.com/openshift/oc/pkg/version"
 )
 
-// NewCmdVersion provides a shim around version for
-// non-client packages that require version information
-func NewCmdVersion(fullName string, versionInfo version.Info, out io.Writer) *cobra.Command {
+type Version struct {
+	kversion.Version
+	OpenShiftVersion string `json:"openshiftVersion,omitempty" yaml:"openshiftVersion,omitempty"`
+}
+
+var (
+	versionLong = templates.LongDesc(`
+		Print the OpenShift client, Kubernetes server, and OpenShift server versions for the current context.
+		Pass --server to print only the Kubernetes and OpenShift server versions.
+		Pass --client to print only the OpenShift client version.`)
+	versionExample = templates.Examples(`
+		# Print the OpenShift client, Kubernetes server, and OpenShift server version information for the current context.
+		%[1]s version
+	    # Print the OpenShift client, Kubernetes server, and OpenShift server version numbers for the current context.
+	    %[1]s version --short
+	    # Print the Kubernetes server and OpenShift server version information for the current context.
+	    %[1]s version --server
+		# Print the OpenShift client version information for the current context.
+		%[1]s version --client`)
+)
+
+type VersionOptions struct {
+	ServerOnly bool
+
+	oClient         configv1client.ClusterOperatorsGetter
+	discoveryClient discovery.CachedDiscoveryInterface
+
+	genericclioptions.IOStreams
+}
+
+func NewVersionOptions(ioStreams genericclioptions.IOStreams) *VersionOptions {
+	return &VersionOptions{
+		IOStreams: ioStreams,
+	}
+}
+
+// NewCmdVersion is copied from upstream NewCmdVersion with added "server" flag
+func NewCmdVersion(fullName string, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	o := NewVersionOptions(ioStreams)
+	ko := kversion.NewVersionOptions(ioStreams)
 	cmd := &cobra.Command{
-		Use:   "version",
-		Short: "Display version",
-		Long:  "Display version",
+		Use:     "version",
+		Short:   i18n.T("Print the client and server version information"),
+		Long:    "Print the client and server version information for the current context",
+		Example: fmt.Sprintf(versionExample, fullName),
 		Run: func(cmd *cobra.Command, args []string) {
-			fmt.Fprintf(out, "%s %v\n", fullName, versionInfo)
+			cmdutil.CheckErr(ko.Validate())
+			cmdutil.CheckErr(o.Complete(f, cmd, ko))
+			cmdutil.CheckErr(o.Run(ko))
 		},
 	}
-
+	cmd.Flags().BoolVar(&ko.ClientOnly, "client", ko.ClientOnly, "Client version only (no server required).")
+	cmd.Flags().BoolVar(&o.ServerOnly, "server", o.ServerOnly, "Server (Kubernetes,OpenShift) versions only.")
```

Comment:
is there a use-case for server-only? I don't see that upstream.

Reply:
I'll remove that, not sure why I added that, looking at the slack conversation, that wasn't mentioned as something we want/need.

Reply:
I don't see a use case for server only, and `--client` should really be `--local` if we're being consistent.

Having both `-short` and `-o` is also inconsistent with our other commands.  `-o version` is probably more appropriate.